### ISSUE TITLE
 FE | Discovery | Profile | Accessibility | Multiple validation errors are not read out to screenreaders

### ISSUE THREAD

**Issue by andaleliz** [department-of-veterans-affairs/va.gov-team#46731]
### User story(ies)

As a user of VA.gov navigating with a screenreader, I need to know when multiple fields have a validation error so I can correct all the problems with the form.

_Example: As a VFS developer, I need to be able to create a custom error message for the XYZ form, so that I can implement the unique error message content that has been approved via the VSP Collaboration Cycle._

### Additional details/notes/links

During a screenreader accessibility audit, we learned that when multiple form fields are in an error state, only the first validation error is announced to a screenreader. Is this a known issue with the form system, or something wrong with our specific implementation? This was seen on [va.gov/profile/contact-information](https://va.gov/profile/contact-information)

Related ticket: #47721 Focus Area

### Tasks:
- [x] Discovery
- [ ] Submit a PR to Design/Forms for feedback and approval

### Acceptance criteria
- [ ] Screenreader users should be able to hear all the validation errors on their forms.

### What type of issue is this? 
- [x] Bug
- [ ] Request new feature
- [ ] Request update to existing feature
- [ ] Documentation change
- [ ] Other (please describe)

**Comment by k80bowman** [comment]
@andaleliz could you please add a link to where we can see this behavior? If you have it, we would also appreciate a link to where this code is being implemented.

**Comment by andaleliz** [comment]
@k80bowman the link is in the description now (it was there previously but not hyperlinked, my bad). You should be able to reproduce it with any LOA3 account on production (or staging) with a screenreader.

I'm not sure exactly what you mean by a link to "where the code is being implemented" - isn't the URL of the page where the code is implemented? If you mean where it lives in the code base, it _might_ be somewhere in [here](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/personalization/profile/components/contact-information), but I am honestly not sure. That's out of my area of expertise.

I don't have more information about this other than what's here already. @AngelaFowler82 did a screenreader walk through of profile back in May, and identified this as an issue.  I think this is problematic because a user can ignore the first error, trigger another, and be unaware that the first error is still there.  Angela, can you confirm that's the expected behavior?

Our team is now going through the backlog and are wondering whether or not this is something our team needs to address or if it's a bug in the form system. Andrés asked me to create this ticket after I asked in platform support, so here I am 😁

**Comment by k80bowman** [comment]
Thank you @andaleliz that is really helpful.

**Comment by andaleliz** [comment]
https://user-images.githubusercontent.com/55992116/189006191-5a96cdd5-ab4f-4c82-9428-d0b1d2adcf63.mov

@cgednalske is a life saver and dug up the recording from the session! You can see what Angela encountered in the video I'm adding with this comment.

**Comment by k80bowman** [comment]
Oh fantastic, thank you!

**Comment by andaleliz** [comment]
Any idea when you might be able to determine if this is our team's issue to fix, or a bug w/ the form system? We're trying to prioritize a bunch of issues related to accessibility and it would be helpful to know if we have to include this. Thanks!

**Comment by k80bowman** [comment]
@andaleliz we don't have a lot of time to devote to forms library issues right now. If your team would like to look into it to see where the problem is, that would be wonderful.

**Comment by andaleliz** [comment]
@SKasimow flagging you for your awareness. This is on our a11y audit list as an item we were not sure whether or not we would fix.

**Comment by SKasimow** [comment]
@andaleliz - yup, it's captured in the Accessibility epic #46756

**Comment by SKasimow** [comment]
@andaleliz - did Forms determine whether it is in their bailiwick to address?

**Comment by andaleliz** [comment]
@SKasimow based on [Katy's comment above](https://github.com/department-of-veterans-affairs/va.gov-team/issues/46731#issuecomment-1243701792) it doesn't seem like they have bandwidth to make that determination.

**Comment by SKasimow** [comment]
@andaleliz - thanks I didn't see that :)

**Comment by andaleliz** [comment]
Per our convo at the end of grooming, another alternative to this could be to explore an error summary pattern.  That would require design + dev work though, and engaging with the experimental design process.

**Comment by SKasimow** [comment]
Thanks @andaleliz

**Comment by andaleliz** [comment]
@SKasimow we should put this in our backlog to further discuss as a team.

**Comment by davidakennedy** [comment]
The Shared Support team can definitely help with this, as @humancompanion-usds said.

To add a bit more color to what's happening in Angela's testing video:

- We use `role="alert"` on our inline error messages to alert screen readers by announcing the error message. This works well in cases where multiple errors aren't a common occurrence. When multiple errors occur, the screen reader has to pick one, and I think that is usually the first or last one in the DOM.
- It would be interesting to know how often multiple errors occur on a page. That may help a team prioritize this work.
- One solution we could explore is using one ARIA live region on a page to announce error messages and customizing it based on the type and number of errors that occur. The downside of this approach is that we'd need  to move focus to the first error message on the page as opposed to a central place that explains all the errors. Also, it may be hard to craft concise, understandable content for multiple errors.
- The error summary is more scalable because you can use it as a companion to inline errors and put focus in a place where all the errors get identified. It also gives you a bit more room for content.

**Comment by SKasimow** [comment]
Hi -  to @davidakennedy point, we will add in a GA custom event to track the the occurrences of multiple errors on address validation. We plan to tackle this in our next sprint (starts next week) and then will review after it is in place for ~ 1 month. Will share findings.

**Comment by adamwhitlock1** [comment]
We did some discovery around the FE for an 'error summary' component, and it turns out there is some existing support for this and a prop exists for `showErrorList` within the schema form component

https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/forms-system/src/js/components/SchemaForm.jsx#L207

The prop is statically set to `false` for all our form instances. This could be made dynamic and would allow an Error List component to display when there are multiple errors in a form.

While experimenting, I set this to true locally just to see what rendered:
![Screen Shot 2022-10-17 at 11.50.21 AM.png](https://images.zenhubusercontent.com/60ddebd504ce2452d0e0fa92/c16e146d-7a3c-46ed-801e-7079c1d6e9b5)

There are of course several issues with this component when it is displayed
- renders the property name, and isn't user friendly in that regard
- since it hasn't been used before, there are design considerations that would need to be addressed for coloration, labeling, spacing and other visual aspects of the list.
- list list actually starts displaying when there is just one error, so that might not be ideal
- no custom error list component could be passed to the schema form to allow some different templating to be used when a custom error list or custom logic is needed in combination with this pattern (this is possible with the newer versions of the core RJSF library, but I don't think we can upgrade that underlying dependency)
- no aria labels or other accessibility considerations have been added to the markup, it is just an h3 and ul for the list of errors that is rendered. 
- some areas that I noticed the error list displaying were actually showing it right away before any form validation happened, and so the form started with it showing all the field errors that weren't really errors yet. 

Example of the error list showing right off the bat:
![Screen Shot 2022-10-17 at 11.53.56 AM.png](https://images.zenhubusercontent.com/60ddebd504ce2452d0e0fa92/2b06788d-cfd9-48ce-8639-2990641dbd80)

**Comment by SKasimow** [comment]
Closing...will open new ticket once we have analytics data
